src/lib/mdns/minimal/ResponseSender.cpp

```diff
@@ -0,0 +1,180 @@
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "ResponseSender.h"
+
+#include "QueryReplyFilter.h"
+
+#include <support/ReturnMacros.h>
+
+#define RETURN_IF_ERROR(err)                                                                                                       \
+    do                                                                                                                             \
+    {                                                                                                                              \
+        if (err)                                                                                                                   \
+        {                                                                                                                          \
+            return;                                                                                                                \
+        }                                                                                                                          \
+    } while (false)
+
+namespace mdns {
+namespace Minimal {
+
+namespace {
+
+constexpr uint16_t kMdnsStandardPort = 5353;
+
+// Restriction for UDP packets:  https://tools.ietf.org/html/rfc1035#section-4.2.1
+//
+//    Messages carried by UDP are restricted to 512 bytes (not counting the IP
+//    or UDP headers).  Longer messages are truncated and the TC bit is set in
+//    the header.
+constexpr uint16_t kPacketSizeBytes = 512;
+
+} // namespace
+
+CHIP_ERROR ResponseSender::Respond(uint32_t messageId, const QueryData & query, const chip::Inet::IPPacketInfo * querySource)
+{
+    mSendError = CHIP_NO_ERROR;
+
+    mCurrentSource    = querySource;
+    mCurrentMessageId = messageId;
+    mSendUnicast      = query.RequestedUnicastAnswer() || (querySource->SrcPort != kMdnsStandardPort);
+    // TODO: at this point we may want to ensure we protect against excessive multicast packet flooding.
+    // According to https://tools.ietf.org/html/rfc6762#section-6  we should multicast at most 1/sec
+    // TBD: do we filter out frequent multicasts or should we switch to unicast in those cases
+
+    // Responder has a stateful 'additional replies required' that is used within the response
+    // loop. 'no additionals required' is set at the start and additionals are marked as the query
+    // reply is built.
+    mResponder->ResetAdditionals();
+
+    mCurrentResourceType = ResourceType::kAnswer; // direct answer
+    QueryReplyFilter filter(query);
+    for (auto it = mResponder->begin(); it != mResponder->end(); it++)
+    {
+        Responder * responder = it->responder;
+
+        if (!filter.Accept(responder->GetQType(), responder->GetQClass(), responder->GetQName()))
```

Comment:
It's common behavior for mdns responders to add any related TXT, SRV and A/AAAA records when it receives a PTR query. This will reduce the packets required for a full query. From the current filter logic it looks like we will not add these records now. Do we plan to support this feature in the future?

Reply:
An example packet with additional TXT/SRV records:

```
Frame 577: 230 bytes on wire (1840 bits), 230 bytes captured (1840 bits) on interface 0
Ethernet II, Src: Google_40:fb:a9 (1c:f2:9a:40:fb:a9), Dst: Apple_6b:f5:ce (88:e9:fe:6b:f5:ce)
Internet Protocol Version 4, Src: 172.23.57.93, Dst: 172.23.57.182
User Datagram Protocol, Src Port: 5353, Dst Port: 5353
Multicast Domain Name System (response)
    Transaction ID: 0x0000
    Flags: 0x8400 Standard query response, No error
    Questions: 0
    Answer RRs: 1
    Authority RRs: 0
    Additional RRs: 3
    Answers
        _meshcop._udp.local: type PTR, class IN, OpenThread-4254-1._meshcop._udp.local
            Name: _meshcop._udp.local
            Type: PTR (domain name PoinTeR) (12)
            .000 0000 0000 0001 = Class: IN (0x0001)
            0... .... .... .... = Cache flush: False
            Time to live: 120
            Data length: 20
            Domain Name: OpenThread-4254-1._meshcop._udp.local
    Additional records
        OpenThread-4254-1._meshcop._udp.local: type TXT, class IN, cache flush
            Name: OpenThread-4254-1._meshcop._udp.local
            Type: TXT (Text strings) (16)
            .000 0000 0000 0001 = Class: IN (0x0001)
            1... .... .... .... = Cache flush: True
            Time to live: 4500
            Data length: 40
            TXT Length: 18
            TXT: nn=OpenThread-4254
            TXT Length: 11
            TXT: xp=��'W'C
            TXT Length: 8
            TXT: tv=1.1.1
        OpenThread-4254-1._meshcop._udp.local: type SRV, class IN, cache flush, priority 0, weight 0, port 49191, target 3dc6e8ca-fc64-d8da-e237-501e453b0079.local
            Service: OpenThread-4254-1
            Protocol: _meshcop
            Name: _udp.local
            Type: SRV (Server Selection) (33)
            .000 0000 0000 0001 = Class: IN (0x0001)
            1... .... .... .... = Cache flush: True
            Time to live: 120
            Data length: 45
            Priority: 0
            Weight: 0
            Port: 49191
            Target: 3dc6e8ca-fc64-d8da-e237-501e453b0079.local
        3dc6e8ca-fc64-d8da-e237-501e453b0079.local: type A, class IN, cache flush, addr 172.23.57.93
            Name: 3dc6e8ca-fc64-d8da-e237-501e453b0079.local
            Type: A (Host Address) (1)
            .000 0000 0000 0001 = Class: IN (0x0001)
            1... .... .... .... = Cache flush: True
            Time to live: 120
            Data length: 4
            Address: 172.23.57.93
    [Unsolicited: True]
```

Reply:
Currently it would return those if the filter is 'ANY' however it would filter them out if the query only asks for PTR for example.

Reply:
Here is how it looks like currently:

```
❯ sudo tshark -j ip -f "udp port 5388" -i any 
Running as user "root" and group "root". This could be dangerous.
Capturing on 'any'
    1 0.000000000   10.0.0.184 5388 224.0.0.251  5353 MDNS 78 Standard query 0x1234 ANY _chip._tcp.local, "QU" question
    2 0.000307879 fe80::224:32ff:fe19:359b 5388 ff02::fb     5353 MDNS 98 Standard query 0x1234 ANY _chip._tcp.local, "QU" question
    3 0.001112022   10.0.0.184 5353 10.0.0.184   5388 MDNS 363 Standard query response 0x1234 PTR chip-mdns-demo._chip._tcp.local TXT SRV 0 0 11097 chip-mdns-demo.local AAAA fe80::224:32ff:fe19:359b A 10.0.0.184
    4 0.035078218   10.0.0.204 5353 10.0.0.184   5388 MDNS 220 Standard query response 0x1234 ANY _chip._tcp.local, "QU" question PTR 2906C908D115D362-8FC7772401CD0696._chip._tcp.local SRV 0 0 22222 Andreis-Mac-mini.local TXT AAAA fe80::1823:ac1e:caee:43c2 A 10.0.0.204
    5 0.035078383 fe80::1823:ac1e:caee:43c2 5353 fe80::224:32ff:fe19:359b 5388 MDNS 240 Standard query response 0x1234 ANY _chip._tcp.local, "QU" question PTR 2906C908D115D362-8FC7772401CD0696._chip._tcp.local SRV 0 0 22222 Andreis-Mac-mini.local TXT AAAA fe80::1823:ac1e:caee:43c2 A 10.0.0.204
^C5 packets captured
```

Reply:
Expanded this is what the current code sends:

```
Frame 5: 363 bytes on wire (2904 bits), 363 bytes captured (2904 bits) on interface any, id 0                                                                                                                                        
Linux cooked capture                                                                                                                                                                                                                 
Internet Protocol Version 4, Src: 10.0.0.184, Dst: 10.0.0.184                                                                                                                                                                        
User Datagram Protocol, Src Port: 5353, Dst Port: 5388                                                                                                                                                                               
Multicast Domain Name System (response)                                                                                                                                                                                              
    Transaction ID: 0x1234                                                                                                                                                                                                           
    Flags: 0x8000 Standard query response, No error                                                                                                                                                                                  
        1... .... .... .... = Response: Message is a response                                                                                                                                                                        
        .000 0... .... .... = Opcode: Standard query (0)                                                                                                                                                                             
        .... .0.. .... .... = Authoritative: Server is not an authority for domain                                                                                                                                                   
        .... ..0. .... .... = Truncated: Message is not truncated                                                                                                                                                                    
        .... ...0 .... .... = Recursion desired: Don't do query recursively                                                                                                                                                          
        .... .... 0... .... = Recursion available: Server can't do recursive queries                                                                                                                                                 
        .... .... .0.. .... = Z: reserved (0)                                                                                                                                                                                        
        .... .... ..0. .... = Answer authenticated: Answer/authority portion was not authenticated by the server                                                                                                                     
        .... .... ...0 .... = Non-authenticated data: Unacceptable                                                                                                                                                                   
        .... .... .... 0000 = Reply code: No error (0)                                                                                                                                                                               
    Questions: 0                                                                                                                                                                                                                     
    Answer RRs: 1                                                                                                                                                                                                                    
    Authority RRs: 0                                                                                                                                                                                                                 
    Additional RRs: 4                                                                                                                                                                                                                
    Answers                                                                                                                                                                                                                          
        _chip._tcp.local: type PTR, class IN, chip-mdns-demo._chip._tcp.local                                                                                                                                                        
            Name: _chip._tcp.local                                                                                                                                                                                                   
            Type: PTR (domain name PoinTeR) (12)                                                                                                                                                                                     
            .000 0000 0000 0001 = Class: IN (0x0001)                                                                                                                                                                                 
            0... .... .... .... = Cache flush: False                                                                                                                                                                                 
            Time to live: 30 (30 seconds)                                                                                                                                                                                            
            Data length: 33                                                                                                                                                                                                          
            Domain Name: chip-mdns-demo._chip._tcp.local                                                                                                                                                                             
    Additional records                                                                                                                                                                                                               
        chip-mdns-demo._chip._tcp.local: type TXT, class IN                                                                                                                                                                          
            Name: chip-mdns-demo._chip._tcp.local                                                                                                                                                                                    
            Type: TXT (Text strings) (16)                                                                                                                                                                                            
            .000 0000 0000 0001 = Class: IN (0x0001)                                                                                                                                                                                 
            0... .... .... .... = Cache flush: False                                                                                                                                                                                 
            Time to live: 10 (10 seconds)                                                                                                                                                                                            
            Data length: 48                                                                                                                                                                                                          
            TXT Length: 9                                                                                                                                                                                                            
            TXT: D0840=yes                                                                                                                                                                                                           
            TXT Length: 10                                                                                                                                                                                                           
            TXT: VP=123+456                                                                                                                                                                                                          
            TXT Length: 4                                                                                                                                                                                                            
            TXT: PH=3                                                                                                                                                                                                                
            TXT Length: 21                                                                                                                                                                                                           
            TXT: OTH=Some text here...
        chip-mdns-demo._chip._tcp.local: type SRV, class IN, priority 0, weight 0, port 11097, target chip-mdns-demo.local
            Service: chip-mdns-demo
            Protocol: _chip
            Name: _tcp.local
            Type: SRV (Server Selection) (33)
            .000 0000 0000 0001 = Class: IN (0x0001)
            0... .... .... .... = Cache flush: False
            Time to live: 30 (30 seconds)
            Data length: 28
            Priority: 0
            Weight: 0
            Port: 11097
            Target: chip-mdns-demo.local
        chip-mdns-demo.local: type AAAA, class IN, addr fe80::224:32ff:fe19:359b
            Name: chip-mdns-demo.local
            Type: AAAA (IPv6 Address) (28)
            .000 0000 0000 0001 = Class: IN (0x0001)
            0... .... .... .... = Cache flush: False
            Time to live: 30 (30 seconds)
            Data length: 16
            AAAA Address: fe80::224:32ff:fe19:359b
        chip-mdns-demo.local: type A, class IN, addr 10.0.0.184
            Name: chip-mdns-demo.local
            Type: A (Host Address) (1)
            .000 0000 0000 0001 = Class: IN (0x0001)
            0... .... .... .... = Cache flush: False
            Time to live: 30 (30 seconds)
            Data length: 4
            Address: 10.0.0.184
    [Unsolicited: True]
```

Reply:
Comparing the records, it looks like we probably want 'cache flush' to be set to true in our implementation as well. For future PRs though (I one pending for replying with the query for legacy operation and also throttling multicast to 1/sec)